File: gtci/09-hashmaps/python3/4-longest-palindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> int:
        freq_map = {}
        for char in s:
            if char not in freq_map:
                freq_map[char] = 0
            freq_map[char] += 1

        length = 0
        at_least_one_odd_char = False
        for char, count in freq_map.items():
            if count % 2 == 0:
                length += count
            else:
                length += count - 1
                at_least_one_odd_char = True

        return length if not at_least_one_odd_char else length + 1


if __name__ == "__main__":
    sol = Solution()

    assert sol.longestPalindrome("abccccdd") == 7
    assert sol.longestPalindrome("a") == 1
    assert sol.longestPalindrome("bb") == 2
    assert sol.longestPalindrome("ccc") == 3
    assert sol.longestPalindrome("abcd") == 1

    print("All test cases passed.")

File: gtci/09-hashmaps/cpp/4-longest-palindrome.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int longestPalindrome(string s) {
        std::unordered_map<char, int> freqMap;

        for (char ch : s) freqMap[ch] += 1;

        int length = 0;
        bool atLeastOneOddChar = false;
        for (const auto& [_, count] : freqMap) {
            if (count % 2 == 0) {
                length += count;
            } else {
                length += count - 1;
                atLeastOneOddChar = true;
            }
        }

        return atLeastOneOddChar ? length + 1 : length;
    }
};

int main() {
    Solution sol;

    assert(sol.longestPalindrome("abccccdd") == 7);
    assert(sol.longestPalindrome("a") == 1);
    assert(sol.longestPalindrome("bb") == 2);
    assert(sol.longestPalindrome("ccc") == 3);
    assert(sol.longestPalindrome("abcd") == 1);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}
